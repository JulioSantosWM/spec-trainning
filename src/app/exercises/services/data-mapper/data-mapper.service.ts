import { Injectable } from "@angular/core";
import { UserDto } from "../../models/user.dto";
import { User } from "../../models/user.model";
import { CreditCardDto } from "../../models/credit-card.dto";
import { StaticParamsService } from "../static-params/static-params.service";
import { DisplayName } from "../../enums/display-name.enum";
import { PrimaryContact } from "../../enums/primary-contact.enum";

@Injectable()
export class DataMapperService {

  constructor(private staticParamsService: StaticParamsService) {}

  toUser(userDto: UserDto): User {
    const user: User = {
      id: userDto.id,
      uid: userDto.uid,
      username: userDto.username,
      name: this.getName(userDto),
      date_of_birth: new Date(userDto.date_of_birth),
      primary_contact: this.getPrimaryContact(userDto),
      avatar: userDto.avatar,
      address: {
        ...userDto.address
      }
    }
    return user;
  }

  toCreditCard(creditCardDto: CreditCardDto): CreditCardDto {
    return creditCardDto;
  }

  private getName(userDto: UserDto): string {
    switch (this.staticParamsService.displayName) {
      case DisplayName.FirstName: 
        return userDto.first_name;
      case DisplayName.LastName:
        return userDto.last_name;
      default:
        return userDto.first_name + ' ' + userDto.last_name;
    }
  }

  private getPrimaryContact(userDto: UserDto): string {
    if (this.staticParamsService.primaryContact === PrimaryContact.Email) {
      return userDto.email;
    }
    return userDto.phone_number;
  }
}