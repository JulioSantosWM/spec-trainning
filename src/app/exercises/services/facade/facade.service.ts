import { Injectable } from "@angular/core";
import { UsersOpenApiService, CreditCardOpenApiService, TemporaryStoreService } from "../../mocks/facade.mock";
import { firstValueFrom, map } from "rxjs";
import { UserDto } from "../../models/user.dto";
import { CreditCardDto } from "../../models/credit-card.dto";

@Injectable()
export class FacadeService {

  constructor(
    private readonly userOpenApiService: UsersOpenApiService,
    private readonly creditCardOpenApiService: CreditCardOpenApiService,
    private readonly temporaryStoreService: TemporaryStoreService,
  ) {
  }

  public getUsers(): Promise<UserDto[]> {
    return firstValueFrom(this.userOpenApiService.getUsers());
  }

  public getUserDetails(uid: string): Promise<UserDto> {
    return firstValueFrom(this.userOpenApiService.getUser(uid));
  }

  public getPrimaryCreditCard(uid: string): Promise<CreditCardDto> {
    return firstValueFrom(this.creditCardOpenApiService.getPrimaryCreditCard(uid));
  }

  public getAvailableCreditCards(uid: string): Promise<CreditCardDto[]> {
    return firstValueFrom(this.creditCardOpenApiService.getAvailableCreditCards(uid));
  }

  public getIsCCExpanded(): Promise<boolean> {
    return firstValueFrom(
      this.temporaryStoreService.getStoredInformation<boolean>('cc_details_expanded')
        .pipe(map(val => !!val))
    );
  }

  public storeIsCCExpanded(isExpanded: boolean): Promise<void> {
    return this.temporaryStoreService.storeInformation('cc_details_expanded', isExpanded);
  }

}