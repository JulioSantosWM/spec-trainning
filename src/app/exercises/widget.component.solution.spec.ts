import { MockBuilder, MockInstance, MockRender, MockedComponentFixture, ngMocks } from "ng-mocks"
import { WidgetComponent } from "./widget.component"
import { DataMapperService } from "./services/data-mapper/data-mapper.service";
import { FacadeService } from "./services/facade/facade.service";
import { GlobalStateService } from "./mocks/global-state.service";
import { createMock } from "@golevelup/ts-jest";
import { User } from "./models/user.model";
import { UserDto } from "./models/user.dto";

describe('WidgetComponent', () => {
  let fixture: MockedComponentFixture<WidgetComponent>;
  let component: WidgetComponent;

  MockInstance.scope();

  beforeEach(() => {
    return MockBuilder(WidgetComponent)
      .mock(FacadeService)
      .mock(DataMapperService)
      .mock(GlobalStateService)
  })

  function setup(): void {
    fixture = MockRender(WidgetComponent);
    component = fixture.point.componentInstance;
  }

  describe('when the component is initializing', () => {
    it('should display a loading spinner', () => {
      setup();

      expect(fixture.nativeElement).toMatchSnapshot();
    })
  })

  describe('when the component has finished loading', () => {
    const userDto = createMock<UserDto>({ uid: 'foo' });
    const user = createMock<User>({ uid: 'bar' });
  
    MockInstance.scope();

    beforeEach(() => {
      MockInstance(FacadeService, 'getIsCCExpanded', jest.fn().mockResolvedValue(true));
      MockInstance(FacadeService, 'getUsers', jest.fn().mockResolvedValue([userDto]));
      MockInstance(DataMapperService, 'toUser', jest.fn().mockReturnValue(user));
    })

    async function loadedSetup(): Promise<void> {
      setup();
      await fixture.whenStable();
      fixture.detectChanges();
    }

    it('should display the table with the data', async() => {
      await loadedSetup();

      expect(fixture.nativeElement).toMatchSnapshot();
      expect(component.rowData).toEqual([user]);
    })

    it('should resize the grid columns when the grid ready', async () => {
      const gridReadyEvent = { api: { sizeColumnsToFit: jest.fn() }}
    
      await loadedSetup();

      ngMocks.output('ag-grid-angular', 'gridReady').emit(gridReadyEvent);

      expect(gridReadyEvent.api.sizeColumnsToFit).toHaveBeenCalledTimes(1);
    })

    describe('when a row is clicked', () => {
      it('should store the toggle value when the credit card is toggled', async () => {
        const spy = MockInstance(FacadeService, 'storeIsCCExpanded', jest.fn());

        await loadedSetup();
        
        ngMocks.output('ag-grid-angular', 'rowClicked').emit({ data: user });
        fixture.detectChanges();
        ngMocks.output('app-w-credit-card-details', 'onExpandCollapseChange').emit(true);

        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith(true);
      })

      it('should update the selected user when a row is clicked', async () => {
        const spy = MockInstance(GlobalStateService, 'updateSelectedUser', jest.fn());

        await loadedSetup();
        
        ngMocks.output('ag-grid-angular', 'rowClicked').emit({ data: user });
        fixture.detectChanges();

        expect(fixture.nativeElement).toMatchSnapshot();
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith(user);
      })
    
      it('should deselect the selected user when the same row is clicked twice', async () => {
        const spy = MockInstance(GlobalStateService, 'updateSelectedUser', jest.fn());
  
        await loadedSetup();
          
        ngMocks.output('ag-grid-angular', 'rowClicked').emit({ data: user });
        fixture.detectChanges();

        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith(user);
  
        ngMocks.output('ag-grid-angular', 'rowClicked').emit({ data: user });
        fixture.detectChanges();
  
        expect(fixture.nativeElement).toMatchSnapshot();
        expect(spy).toHaveBeenCalledTimes(2);
        expect(spy).toHaveBeenCalledWith(null);
      })
    })
  })
})