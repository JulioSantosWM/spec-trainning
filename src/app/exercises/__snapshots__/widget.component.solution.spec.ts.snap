// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetComponent when the component has finished loading should display the table with the data 1`] = `
<div
  id="root1"
>
  <app-widget>
    <mat-card>
      <mat-card-header>
        <mat-card-title>
          User Details Dashboard
        </mat-card-title>
        
      </mat-card-header>
      <mat-card-content>
        <br />
        <div
          class="main-container"
        >
          <ag-grid-angular
            class="ag-theme-alpine"
            style="width: 1000px; height: 600px;"
          />
          
        </div>
        
      </mat-card-content>
      
    </mat-card>
    
  </app-widget>
</div>
`;

exports[`WidgetComponent when the component has finished loading when a row is clicked should deselect the selected user when the same row is clicked twice 1`] = `
<div
  id="root5"
>
  <app-widget>
    <mat-card>
      <mat-card-header>
        <mat-card-title>
          User Details Dashboard
        </mat-card-title>
        
      </mat-card-header>
      <mat-card-content>
        <br />
        <div
          class="main-container"
        >
          <ag-grid-angular
            class="ag-theme-alpine"
            style="width: 1000px; height: 600px;"
          />
          
        </div>
        
      </mat-card-content>
      
    </mat-card>
    
  </app-widget>
</div>
`;

exports[`WidgetComponent when the component has finished loading when a row is clicked should update the selected user when a row is clicked 1`] = `
<div
  id="root4"
>
  <app-widget>
    <mat-card>
      <mat-card-header>
        <mat-card-title>
          User Details Dashboard
        </mat-card-title>
        
      </mat-card-header>
      <mat-card-content>
        <br />
        <div
          class="main-container"
        >
          <ag-grid-angular
            class="ag-theme-alpine"
            style="width: 1000px; height: 600px;"
          />
          <div
            class="details-container"
          >
            <app-w-user-focus />
            <app-w-credit-card-details
              class="credit-card-details"
            />
          </div>
          
        </div>
        
      </mat-card-content>
      
    </mat-card>
    
  </app-widget>
</div>
`;

exports[`WidgetComponent when the component is initializing should display a loading spinner 1`] = `
<div
  id="root0"
>
  <app-widget>
    <mat-card>
      <mat-card-header>
        <mat-card-title>
          User Details Dashboard
        </mat-card-title>
        
      </mat-card-header>
      <mat-card-content>
        <br />
        <div
          class="spinner-container"
        >
          <mat-spinner />
        </div>
        
      </mat-card-content>
      
    </mat-card>
    
  </app-widget>
</div>
`;
